Network/SerializationHooks.cs

```diff
@@ -37,57 +37,52 @@ public static void Uninitialize()
         _deserializeDetour?.Dispose();
     }
 
+
     [UnmanagedFunctionPointer(CallingConvention.StdCall)]
-    public unsafe delegate void SerializeEvent(IntPtr entityManager, UInt64 networkEventType, UInt64 netBufferOut, IntPtr entity);
+    public unsafe delegate void SerializeEvent(IntPtr entityManager, IntPtr networkEventType, IntPtr netBufferOut, IntPtr entity);
 
     public static SerializeEvent? SerializeOriginal;
 
-    public unsafe static void SerializeHook(IntPtr entityManager, UInt64 networkEventType, UInt64 netBufferOut, IntPtr entity)
+    public unsafe static void SerializeHook(IntPtr entityManager, IntPtr networkEventType, IntPtr netBuffer, IntPtr entity)
     {
         var eventType = *(NetworkEventType*)&networkEventType;
 
         // if this is not a custom event, just call the original function
         if (eventType.EventId != SerializationHooks.WETSTONE_NETWORK_EVENT_ID)
         {
-            SerializeOriginal!(entityManager, networkEventType, netBufferOut, entity);
+            SerializeOriginal!(entityManager, networkEventType, netBuffer, entity);
             return;
         }
 
-        // we need to adjust by 0x10 here because the managed proxy that Il2CppUnhollower
-        // generates expects a boxed struct, whereas we have a pointer to an unboxed one.
-        // subtracting 0x10 is exactly two pointers, which allows us to fake a boxed struct
-        // with the same underlying data
-        var netBuffer = new NetBufferOut(new IntPtr((long)(netBufferOut - 0x10)));
+        ref NetBufferOut netBufferOut = ref *(NetBufferOut*)netBuffer;
 
         // extract the custom network event
         var realEntity = *(Entity*)&entity;
         var data = (CustomNetworkEvent)VWorld.Server.EntityManager.GetComponentObject<Il2CppSystem.Object>(realEntity, CustomNetworkEvent.ComponentType);
 
         // write out the event ID and the data
-        netBuffer.Write((uint)SerializationHooks.WETSTONE_NETWORK_EVENT_ID);
-        data.Serialize(netBuffer);
+        netBufferOut.Write((uint)SerializationHooks.WETSTONE_NETWORK_EVENT_ID);
+        data.Serialize(ref netBufferOut);
     }
 
     // --------------------------------------------------------------------------------------
 
     [UnmanagedFunctionPointer(CallingConvention.StdCall)]
-    public unsafe delegate void DeserializeEvent(IntPtr commandBuffer, IntPtr netBuffer, DeserializeNetworkEventParams eventParams);
+    public unsafe delegate void DeserializeEvent(IntPtr entityManager, IntPtr commandBuffer, IntPtr netBuffer, DeserializeNetworkEventParams eventParams);
 
     public static DeserializeEvent? DeserializeOriginal;
 
-    public unsafe static void DeserializeHook(IntPtr commandBuffer, IntPtr netBuffer, DeserializeNetworkEventParams eventParams)
+    public unsafe static void DeserializeHook(IntPtr entityManager, IntPtr commandBuffer, IntPtr netBuffer, DeserializeNetworkEventParams eventParams)
```

Comment:
Taking these parameters as native integers is redundant, you end up turning them into the relevant ref structs anyway. Take them in as ref parameters instead.
```suggestion
    public unsafe static void DeserializeHook(IntPtr entityManager, IntPtr commandBuffer, ref NetBufferIn netBuffer, DeserializeNetworkEventParams eventParams)
```